# PyTorch Tensor Basics

In [2]:
#Imports
import torch
import torch.nn as nn


## Introduction To Tensors

- A tensor is a main data structure in PyTorch.
- It is similar to NumPy array, but with two important advantages:
  - tensors can run on GPU
  - tensors support **automatic differentiation** using PyTorch Autograd.

Tensors are used to represent:

- Input data
- model parameters
- intermediate activations
- gradients
- predictions
- losses

Common tensor dimensions:


In [3]:
scalar = torch.tensor(5)              # 0D tensor
vector = torch.tensor([1, 2, 3])      # 1D tensor
matrix = torch.tensor([[1, 2], [3, 4]])  # 2D tensor

In deep learning, tensor shape is extremely important.

```
# Image batch
(batch_size, channels, height, width)

# Text batch
(batch_size, sequence_length)

# Tabular data
(batch_size, num_features)
```

### Creating Tensors

#### From Python Lists

In [5]:
x = torch.tensor([1, 2, 3])
y = torch.tensor([[1, 2], [3, 4]])
print(f"Shape of x: {x.shape}")  # Output: torch.Size([3])
print(f"Shape of y: {y.shape}")  # Output: torch.Size([2, 2])

Shape of x: torch.Size([3])
Shape of y: torch.Size([2, 2])


#### Zeros and Ones

In [6]:
zeros = torch.zeros(3, 4)
ones = torch.ones(2, 3)
print(f"Zeros tensor:\n{zeros}")
print(f"Ones tensor:\n{ones}")

Zeros tensor:
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])
Ones tensor:
tensor([[1., 1., 1.],
        [1., 1., 1.]])


#### Random Tensors

- `torch.randn()` is commonly used for initializing weights.

In [8]:
rand = torch.rand(3, 4)      # values between 0 and 1
randn = torch.randn(3, 4)    # values from normal distribution
randint = torch.randint(0, 10, (3, 4))  # random integers between 0 and 9
print(f"Random tensor (0 to 1):\n{rand}")
print(f"Random tensor (normal distribution):\n{randn}")
print(f"Random integer tensor (0 to 9):\n{randint}")


Random tensor (0 to 1):
tensor([[0.0775, 0.6819, 0.6673, 0.1884],
        [0.4632, 0.6549, 0.3775, 0.4148],
        [0.5866, 0.1628, 0.9829, 0.1698]])
Random tensor (normal distribution):
tensor([[ 0.7905,  0.4488,  1.2063,  1.3174],
        [-0.1427,  0.4108, -0.4196,  1.1496],
        [ 1.6881,  0.8962,  0.4174,  0.0651]])
Random integer tensor (0 to 9):
tensor([[3, 1, 2, 9],
        [0, 8, 0, 0],
        [6, 1, 7, 9]])


#### Range Tensors

In [9]:
x = torch.arange(0, 10)
y = torch.arange(0, 10, 2)
z = torch.linspace(0, 1, steps=5)
print(f"Arange (0 to 9):\n{x}")
print(f"Arange (0 to 8, step 2):\n{y}")
print(f"Linspace (0 to 1, 5 steps):\n{z}")


Arange (0 to 9):
tensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
Arange (0 to 8, step 2):
tensor([0, 2, 4, 6, 8])
Linspace (0 to 1, 5 steps):
tensor([0.0000, 0.2500, 0.5000, 0.7500, 1.0000])


#### Creating Tensors Like Another Tensor

In [10]:
x = torch.ones(3, 4)

zeros_like = torch.zeros_like(x)
ones_like = torch.ones_like(x)
rand_like = torch.rand_like(x)
print(f"Zeros like x:\n{zeros_like}")
print(f"Ones like x:\n{ones_like}")
print(f"Random like x:\n{rand_like}")


Zeros like x:
tensor([[0., 0., 0., 0.],
        [0., 0., 0., 0.],
        [0., 0., 0., 0.]])
Ones like x:
tensor([[1., 1., 1., 1.],
        [1., 1., 1., 1.],
        [1., 1., 1., 1.]])
Random like x:
tensor([[0.7766, 0.8123, 0.2487, 0.8434],
        [0.8642, 0.1482, 0.5794, 0.0895],
        [0.7143, 0.5613, 0.1217, 0.6898]])


### Tensor data types

- Every Tensor has a data type, also called `dtype`.

In [11]:
x = torch.tensor([1, 2, 3])
x.dtype

torch.int64

Common PyTorch data types:

|Data Type|Meaning|Common Use|
|---|---|---|
|`torch.float32`|32-bit floating point|Default for neural networks|
|`torch.float34`|64-bit floating point|High Precision numerical work|
|`torch.float16`|16-bit floating point|Mixed precision training|
|`torch.bfloat16`|brain floating point|Efficient deep learning on modern hardware|
|`torch.int64`|64-bit integer|Class labels, indexing|
|`torch.int32`|32-bit integer|integer data|
|`torch.bool`|Boolean|Masks and conditions|

#### Specifying dtype

In [15]:
x = torch.tensor([1, 2, 3], dtype=torch.float32)

print(f"Tensor x:\n{x}")
print(f"Data type of x: {x.dtype}")

Tensor x:
tensor([1., 2., 3.])
Data type of x: torch.float32


#### Converting dtype

In [16]:
x = x.float()
print(f"Tensor x (float):\n{x}")

x = x.long()
print(f"Tensor x (long):\n{x}")

x = x.bool()
print(f"Tensor x (bool):\n{x}")

Tensor x (float):
tensor([1., 2., 3.])
Tensor x (long):
tensor([1, 2, 3])
Tensor x (bool):
tensor([True, True, True])


In [17]:
x = x.to(torch.float32)
print(f"Tensor x (float):\n{x}")

Tensor x (float):
tensor([1., 1., 1.])


> Important Rule:
>
> - `nn.CrossEntropyLoss()` expects
>   - models outputs: `float32` logits
>   - class labels: `int64`, also called `long`

In [18]:
labels = torch.tensor([0, 2, 1], dtype=torch.long)
print(f"Labels tensor:\n{labels}")
print(f"Data type of labels: {labels.dtype}")

Labels tensor:
tensor([0, 2, 1])
Data type of labels: torch.int64


#### Useful shape-related attributes:

- `x.shape` :  # shape of tensor
- `x.ndim` : # number of dimensions
- `x.size()` : # same as shape
- `x.numel()` : # total number of elements

In [22]:
x = torch.randint(0, 10, (3, 4))

print(f"Tensor x:\n{x}")
print(f"Shape of x: {x.shape}")
print(f"Number of dimensions of x: {x.ndim}")
print(f"Number of elements in x: {x.numel()}")


Tensor x:
tensor([[0, 2, 8, 0],
        [6, 9, 7, 2],
        [8, 8, 0, 0]])
Shape of x: torch.Size([3, 4])
Number of dimensions of x: 2
Number of elements in x: 12


### Tensor shapes

- The shape of a tensor tells us the size of each dimension.

In [19]:
x = torch.randn(2, 3, 4)
x.shape

torch.Size([2, 3, 4])

#### Common Shape Patterns

For tabular data:
```
(batch_size, num_features)
```

For images in PyTorch
```
(batch_size, channels, height, width)
```

Example:

In [25]:
images = torch.randn(32, 3, 224, 224)
print(f"Shape of images: {images.shape}")

Shape of images: torch.Size([32, 3, 224, 224])


This means:

- 32 images
- 3 color channels
- height 224
- width 224

### Reshaping tensors

Reshaping changes the shape of a tensor without changing its data.

#### `reshape()`


In [26]:
x = torch.arange(12)
print(f"Original tensor x:\n{x}")
x = x.reshape(3, 4)
print(f"Reshaped tensor x:\n{x}")

Original tensor x:
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])
Reshaped tensor x:
tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])


#### `view()`

- view() is faster in some cases, but requires tensor to be contiguous in memory.
- `reshape()` is safer because it works even when the tensor is not contiguous.

In [28]:
x = torch.arange(12)
print(f"Original tensor x:\n{x}")
x = x.view(3, 4)
print(f"Viewed tensor x:\n{x}")

Original tensor x:
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])
Viewed tensor x:
tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])


#### Flattening

In [30]:
x = torch.randn(2, 3, 4)
x_flat = x.reshape(-1)
print(f"Shape of original x: {x.shape}")
print(f"Shape of flattened x: {x_flat.shape}")

Shape of original x: torch.Size([2, 3, 4])
Shape of flattened x: torch.Size([24])


In [31]:
x = torch.randn(32, 3, 28, 28)
print(f"Shape of x: {x.shape}")
x = x.reshape(32, -1)
print(f"Shape of x after flattening: {x.shape}")

Shape of x: torch.Size([32, 3, 28, 28])
Shape of x after flattening: torch.Size([32, 2352])


The -1 means PyTorch automatically calculates that dimension.

#### Adding Dimension

In [32]:
x = torch.tensor([1, 2, 3])
print(f"Shape of x: {x.shape}")
x = x.unsqueeze(0)
print(f"Shape of x after unsqueeze: {x.shape}")

Shape of x: torch.Size([3])
Shape of x after unsqueeze: torch.Size([1, 3])


Useful for adding a batch dimension.

#### Removing Dimensions

- Removes dimensions of size 1.

In [33]:
x = torch.randn(1, 3, 1, 5)
print(f"Original shape of x: {x.shape}")
x = x.squeeze()
print(f"Shape of x after squeeze: {x.shape}")

Original shape of x: torch.Size([1, 3, 1, 5])
Shape of x after squeeze: torch.Size([3, 5])


> Be Careful
>
> - `x.squeeze()` can accidently remove the batch dimension if batch size is 1.

#### Rearranging dimensions

In [35]:
x = torch.randn(32, 224, 224, 3)
x = x.permute(0, 3, 1, 2)
print(f"Shape of x after permute: {x.shape}")

Shape of x after permute: torch.Size([32, 3, 224, 224])


### Indexing and Slicing

- Indexing is used to access specific elements or parts of a tensor.

In [36]:
x = torch.tensor([10, 20, 30, 40])
x[0]

tensor(10)

In [37]:
x = torch.tensor([[1, 2, 3],
                  [4, 5, 6]])

x[0, 1]


tensor(2)

In [38]:
# Slicing

x = torch.arange(10)

x[2:6]

tensor([2, 3, 4, 5])

#### Row and Column Selection

In [39]:
x = torch.tensor([[1, 2, 3],
                  [4, 5, 6],
                  [7, 8, 9]])


In [40]:
x[0]  # first row

tensor([1, 2, 3])

In [41]:
# first column
x[:, 0]

tensor([1, 4, 7])

In [42]:
# First two rows
x[:2]

tensor([[1, 2, 3],
        [4, 5, 6]])

In [43]:
# last column
x[:, -1]

tensor([3, 6, 9])

#### Boolean Indexing

- Useful for filtering values

In [44]:
x = torch.tensor([1, 2, 3, 4, 5])

x[x > 3]

tensor([4, 5])

#### Masking

- Masks are common in NLP, attention models, and loss filtering.

In [45]:
x = torch.tensor([1, 2, 3, 4, 5])
mask = x % 2 == 0

x[mask]

tensor([2, 4])

#### Advanced indexing

- Useful when selecting specific samples, labels, or embeddings

In [46]:
x = torch.tensor([10, 20, 30, 40])
indices = torch.tensor([0, 2])

x[indices]


tensor([10, 30])

Key Notes

- Always check tensor shapes when debugging neural networks.
- Most PyTorch shape errors come from incorrect dimensions.
- Pytorch image format is usually:
  - (batch, channels, height, width)
- Class labels for classification usually need `torch.long`
- Neural networks inputs and weights usually use `torch.float32`
- Use `reshape()` when unsure; use `view()` only when you understand memory contiguilty.
- Use `unsqueeze()` to add a batch dimension.

## Tensor Operations

- Tensors operations are the core of deep learning in PyTorch.

### Arithmetic Operations

- Arithmetic operations are basic mathematical operations performed on tensors.
- Common arithmetic operations:

```py
A + B   # addition
A - B   # subtraction
A * B   # multiplication
A / B   # division
A ** 2  # power
```

In [47]:
import torch

a = torch.tensor([1, 2, 3])
b = torch.tensor([4, 5, 6])

print(a + b)  # tensor([5, 7, 9])
print(a * b)  # tensor([4, 10, 18])

tensor([5, 7, 9])
tensor([ 4, 10, 18])


Arithmetic operations can be performed:

- Tensor with tensor
- Tensor with scalar
- Tensor with broadcastable tensor

In [48]:
a = torch.tensor([1, 2, 3])
print(a + 10)  # tensor([11, 12, 13])

tensor([11, 12, 13])


### Broadcasting

- Broadcasting allows operations between tensors of different shapes.
- Here, b is automatically expanded to match the shape of a.

Broadcasting is important because it avoids writing loops and makes computation faster.

Broadcasting rules:

- Dimensions are compared from right to left.
- Dimensions must either be equal, or one of them must be 1.
- If shapes are incompatible, an error occurs.

```py
(2, 3) and (3,)  -> valid
(4, 1) and (4, 5) -> valid
(2, 3) and (2,) -> invalid
```

In [49]:
a = torch.tensor([[1, 2, 3],
                  [4, 5, 6]])

b = torch.tensor([10, 20, 30])

print(a + b)

tensor([[11, 22, 33],
        [14, 25, 36]])


### Matrix Multiplication

- Matrix multiplication is one of the most important tensor operations in neural networks.
- It is different from element-wise multiplication.
- Matrix multiplication shape rule:
  - The inner dimensions must match.
```py
A shape = (m, n)
B shape = (n, p)

A @ B shape = (m, p)
```

Matrix multiplication is used in:

```
Dense layers
Attention mechanisms
Embedding projections
Convolution internally
Transformer models
Backpropagation
```

In [ ]:
# Element-Wise Multiplication
A  = torch.tensor([[1, 2],
                   [3, 4]])
B = torch.tensor([[5, 6],
                   [7, 8]])

A * B

tensor([[ 5, 12],
        [21, 32]])

In [ ]:
# Matrix Multiplication
A @ B # OR torch.matmul(A, B)

tensor([[19, 22],
        [43, 50]])

In neural networks, matrix multiplication is used for fully connected layers:

```py
output = input @ weights + bias
```

if
```py
input shape   = (batch_size, input_features)
weights shape = (input_features, output_features)
bias shape    = (output_features)
```

then
```py
output shape = (batch_size, output_features)
```

#### Dot Product

- The dot product is a special case of matrix multiplication between two vectors.
- In deep learning, dot products are used to compute similarity and weighted sums.



In [52]:
a = torch.tensor([1, 2, 3])
b = torch.tensor([4, 5, 6])

print(torch.dot(a, b))


tensor(32)


### Aggregation Operations

- Aggregation operations reduce multiple tensor values into fewer values.

Common aggregation operations:

```py
torch.sum()
torch.mean()
torch.max()
torch.min()
torch.argmax()
torch.argmin()
torch.prod()
torch.std()
torch.var()
```

In [54]:
x = torch.tensor([[1, 2, 3],
                  [4, 5, 6]])

print(torch.sum(x))   # tensor(21)
print(torch.mean(x.float()))  # tensor(3.5000)
print(torch.max(x))   # tensor(6)
print(torch.min(x))   # tensor(1)
print(torch.std(x.float()))  # tensor(1.8708)
print(torch.argmax(x))  # tensor(5) - index of max value in flattened tensor
print(torch.argmin(x))  # tensor(0) - index of min value in flattened tensor
print(torch.unique(x))  # tensor([1, 2, 3, 4, 5, 6])
print(torch.nonzero(x > 3))  # tensor([[1, 0], [1, 1], [1, 2]]) - indices of elements greater than 3
print(torch.where(x > 3, x, torch.tensor(0)))  # tensor([[0, 0, 0], [4, 5, 6]]) - replace elements > 3 with their value, else 0
print(torch.any(x > 5))  # tensor(True) - checks if any element is greater than 5
print(torch.all(x > 0))  # tensor(True) - checks if all elements are
print(torch.all(x > 3))  # tensor(False) - checks if all elements are greater than 3


tensor(21)
tensor(3.5000)
tensor(6)
tensor(1)
tensor(1.8708)
tensor(5)
tensor(0)
tensor([1, 2, 3, 4, 5, 6])
tensor([[1, 0],
        [1, 1],
        [1, 2]])
tensor([[0, 0, 0],
        [4, 5, 6]])
tensor(True)
tensor(True)
tensor(False)


Aggregation can be done over the entire tensor or along a specific dimension.

In [55]:
x = torch.tensor([[1, 2, 3],
                  [4, 5, 6]])

print(torch.sum(x, dim=0))
print(torch.sum(x, dim=1))


tensor([5, 7, 9])
tensor([ 6, 15])


In deep learning, aggregation operations are used in:

```
Loss functions
Accuracy calculation
Pooling layers
Batch statistics
Gradient reduction
Attention score normalization
Global average pooling
```

### Element-wise Operations

Element-wise operations apply the same operation to each element of a tensor independently.

In [56]:
a = torch.tensor([1, 2, 3])
b = torch.tensor([4, 5, 6])

print(a * b)
print(a.mul(b))  # element-wise multiplication
print(torch.mul(a, b))  # element-wise multiplication

tensor([ 4, 10, 18])
tensor([ 4, 10, 18])
tensor([ 4, 10, 18])


Common element-wise operations:

```py
torch.add()
torch.sub()
torch.mul()
torch.div()
torch.exp()
torch.log()
torch.sqrt()
torch.abs()
torch.sin()
torch.cos()
torch.relu()
torch.sigmoid()
torch.tanh()
```

Element-wise operations are heavily used in activation functions.


### In-place Operations

- In-place operations modify the original tensor directly instead of creating a new tensor.
- In PyTorch, in-place operations usually end with an underscore _.


In [59]:
x = torch.tensor([[1, 2, 3],
                  [4, 5, 6]])

x = x.to(torch.float32)

print(x.add_(5))
print(x.sub_(2))
print(x.mul_(3))
print(x.div_(2))
print(x.relu_())

tensor([[ 6.,  7.,  8.],
        [ 9., 10., 11.]])
tensor([[4., 5., 6.],
        [7., 8., 9.]])
tensor([[12., 15., 18.],
        [21., 24., 27.]])
tensor([[ 6.0000,  7.5000,  9.0000],
        [10.5000, 12.0000, 13.5000]])
tensor([[ 6.0000,  7.5000,  9.0000],
        [10.5000, 12.0000, 13.5000]])


In [60]:
x = torch.tensor([1, 2, 3])
x.add_(10)

print(x)


tensor([11, 12, 13])


- In-place operations save memory because they do not create a new tensor.
- However, they can be dangerous in deep learning because they may interfere with automatic differentiation.
- PyTorch tracks operations for backpropagation. If a tensor needed for gradient calculation is modified in-place, PyTorch may raise an error or compute incorrect gradients.
- Important exam point:
  - In-place operations are memory efficient but should be used carefully when gradients are involved.

## CPU and GPU Computation

- Deep learning models perform many mathematical operations on tensors. These operations can run on a CPU or a GPU.
- A CPU is general-purpose and good for many types of tasks.
- A GPU is designed for parallel computation, so it is much faster for large tensor operations, matrix multiplication, image processing, and neural network training.
- In PyTorch, the place where a tensor is stored and computed is called a device.

### Devices in PyTorch

A PyTorch tensor can be stored on different devices.
  
```py
"cpu"
"cuda"
"cuda:0"
"cuda:1"
```

By default, tensors are created on the CPU unless specified otherwise.

In [61]:
import torch

x = torch.tensor([1, 2, 3])
print(x.device)


cpu


To create a tensor directly on a device:

In [62]:
x = torch.tensor([1, 2, 3], device="cpu")
print(x.device)

cpu


In [64]:
# x = torch.tensor([1, 2, 3], device="cuda")
# print(x.device)

In [65]:
# x = torch.tensor([1, 2, 3], device="cuda:0")
# print(x.device)

Here:

- cuda:0 means first GPU
- cuda:1 means second GPU

### Checking CUDA Availability

- UDA is NVIDIA’s parallel computing platform used by PyTorch to run tensor operations on NVIDIA GPUs.
- Before using GPU, check whether CUDA is available:

In [67]:
import torch

print(torch.cuda.is_available())

False


In [69]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(device)

cpu


### Moving tensors to GPU

To move a tensor from CPU to GPU, use:

In [72]:
# x = x.to("cuda")

In [73]:
# x = x.cuda()

In [74]:
x = x.to(device)

In [75]:
import torch

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

x = torch.tensor([1.0, 2.0, 3.0])
x = x.to(device)

print(x.device)

cpu


Important point:

- to() does not always modify the original tensor in-place. It returns a tensor on the requested device.

```
x.to(device) # incorrect

x = x.to(device)  # correct
```

This is needed when converting tensors to NumPy arrays:

In [76]:
x_numpy = x.cpu().numpy()
print(x_numpy)

[1. 2. 3.]


A CUDA tensor cannot be directly converted to NumPy.

### Moving Models to GPU

- Neural network models must also be moved to the same device as the data.

In [77]:
import torch
import torch.nn as nn

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = nn.Linear(10, 2)
model = model.to(device)

In [79]:
inputs = torch.randn(5, 10)
labels = torch.tensor([0, 1, 0, 1, 0])

In [80]:
# Training example
inputs = inputs.to(device)
labels = labels.to(device)

outputs = model(inputs)


Important rule:

> The model and the input tensors must be on the same device.

### Common Device Mismatch errors

- A device mismatch error happens when tensors involved in the same operation are on different devices.

Common Causes of device mismatch:

1. Model is on GPU but input is on CPU.
2. Labels are left on CPU while outputs are on GPU.
3. New tensors are created inside training loop without specifying device.
4. Hidden states in RNNs are created on CPU while model is on GPU.
5. NumPy arrays are converted to tensors but remain on CPU.


## Reproduccibility

Reproducibility means getting the same or nearly same results when running the same code again.

In deep learning, reproducibility is important for:

- Debugging
- Comparing models
- Scientific experiments
- Examining whether improvements are real
- Sharing results

Deep learning results can vary because many operations involve randomness.

### Random seeds

A random seed fixes the starting point of a random number generator.

If the same seed is used, the same sequence of random numbers is generated.

Example:


In [83]:
import torch

torch.manual_seed(42)

x = torch.rand(3)
print(x)

tensor([0.8823, 0.9150, 0.3829])


Set seeds for PyTorch:

In [84]:
torch.manual_seed(42)

For CUDA:

In [85]:
torch.cuda.manual_seed(42)
torch.cuda.manual_seed_all(42)

For Python random module:

In [86]:
import random
random.seed(42)

For NumPy:

In [87]:
import numpy as np
np.random.seed(42)

### Deterministic Behavior

Deterministic behavior means the same input and same code produce the same output every time.

In PyTorch, some GPU operations may be non-deterministic because they use optimized parallel algorithms.

To request deterministic behavior:

In [88]:
torch.use_deterministic_algorithms(True)

Important exam point:

> Deterministic settings can make training slower because PyTorch may avoid faster non-deterministic algorithms.

### Why Results May Still Differ

Even after setting seeds, results may still differ slightly.

Reasons include:

1. GPU non-determinism
   - Some CUDA operations use parallel execution where operation order may vary.
2. Floating-point precision
   - Floating-point arithmetic is approximate.

Example:

```
(a + b) + c may not equal a + (b + c)
```

3. Different hardware
    - CPU, GPU, and different GPU models may produce slightly different results.
4. Different PyTorch/CUDA/cuDNN versions
    - Different library versions may use different algorithms.
5. Non-deterministic operations
   - Some operations do not have deterministic implementations on all devices.
6. Multi-threading
   - Parallel CPU/GPU operations may execute in different orders.
7. DataLoader workers
    - Multiple workers can introduce randomness if not seeded properly.

# Automatic Differentiation

Automatic differentiation is the technique deep learning libraries use to automatically calculate gradients.

In neural networks, we need gradients to know:
> How should each weight change so that the loss becomes smaller?


Instead of manually differentiating every equation, PyTorch uses autograd to track tensor operations and calculate derivatives automatically.

## What is Autograd?

### Autograd

- Autograd means automatic gradient calculation.
- PyTorch’s autograd system records operations performed on tensors and uses them to compute gradients during backpropagation.

In [89]:
import torch

x = torch.tensor(2.0, requires_grad=True)

y = x ** 2

y.backward()

print(x.grad)

tensor(4.)


Explanation

```
y = x²
dy/dx = 2x
when x = 2, dy/dx = 4
```

PyTorch automatically calculated the derivative.

### Computational Graphs

- A computational graph is a graph that represents mathematical operations.
- It shows how inputs are transformed into outputs.

Example:

In [90]:
x = torch.tensor(2.0, requires_grad=True)
y = x * 3
z = y + 4
loss = z ** 2

Mathematically:

```
y = 3x
z = y + 4
loss = z²
```

Graph:

```
x → multiply by 3 → y → add 4 → z → square → loss
```

Each operation becomes a node in the graph.

The graph stores information needed to calculate gradients later.

Important terms:

|Term|Meaning|
|---|---|
|Node|Operation or tensor in the graph|
|Edge|Connection showing dependency|
|Input tensor|Starting tensor|
|Output tensor|Result tensor|
|Loss|Final scalar value usually used for backpropagation|

In PyTorch, computational graphs are usually dynamic.

This means the graph is created while the code runs.

This is called:
```
define-by-run
```


1. Neural Network Foundations
1.1 What Is a Neural Network?
Biological intuition vs mathematical model
Neurons, layers, weights, biases
Input layer, hidden layers, output layer
Shallow vs deep neural networks
1.2 Why Deep Learning Works
Representation learning
Hierarchical feature learning
Why depth matters
When deep learning is useful
1.3 Core Mathematical Objects
Scalars, vectors, matrices, tensors
Tensor shapes and dimensions
Broadcasting
Dot products and matrix multiplication
1.4 Linear Models as Neural Networks
Linear regression as a single neuron
Logistic regression as a simple neural network
From linear models to multilayer networks
2. PyTorch Tensor Basics
2.1 Introduction to Tensors
Creating tensors
Tensor data types
Tensor shapes
Reshaping tensors
Indexing and slicing
2.2 Tensor Operations
Arithmetic operations
Matrix multiplication
Aggregation operations
Element-wise operations
In-place operations
2.3 CPU and GPU Computation
Devices in PyTorch
Moving tensors to GPU
Checking CUDA availability
Common device mismatch errors
2.4 Reproducibility
Random seeds
Deterministic behavior
Why results may still differ
3. Automatic Differentiation
3.1 What Is Autograd?
Computational graphs
Forward pass
Backward pass
Gradients
3.2 Gradients in PyTorch
requires_grad
.grad
.backward()
torch.no_grad()
Detaching tensors
3.3 Understanding Gradient Flow
Chain rule
Leaf tensors
Gradient accumulation
Zeroing gradients
4. Building Neural Networks in PyTorch
4.1 The nn.Module Class
Why use nn.Module
Defining custom models
__init__() and forward()
Parameters and buffers
4.2 Layers in PyTorch
nn.Linear
Activation layers
Dropout
Batch normalization
Flattening and reshaping layers
4.3 Forward Pass
What happens during prediction
Model inputs and outputs
Shape tracking through layers
4.4 Model Parameters
Weights and biases
Accessing model parameters
Counting trainable parameters
Freezing parameters
5. Activation Functions
5.1 Why Activation Functions Are Needed
Non-linearity
Why stacked linear layers are still linear
5.2 Common Activation Functions
Sigmoid
Tanh
ReLU
Leaky ReLU
GELU
Softmax
5.3 Choosing Activation Functions
Hidden layer activations
Output layer activations
Problems with saturation
Dead ReLU problem
6. Loss Functions
6.1 What Is a Loss Function?
Objective function
Training signal
Loss vs metric
6.2 Regression Losses
Mean Squared Error
Mean Absolute Error
Huber loss
6.3 Classification Losses
Binary Cross Entropy
Cross Entropy Loss
Negative Log Likelihood Loss
6.4 Matching Outputs and Losses
Regression outputs
Binary classification outputs
Multiclass classification outputs
Logits vs probabilities
7. Optimization
7.1 Gradient Descent
Intuition
Learning rate
Parameter updates
7.2 Optimizers in PyTorch
SGD
SGD with momentum
RMSProp
Adam
AdamW
7.3 Learning Rate
Too high vs too low
Learning rate schedules
Warmup
Weight decay
7.4 Optimization Problems
Local minima
Saddle points
Vanishing gradients
Exploding gradients
8. Training Neural Networks
8.1 The Training Loop
Forward pass
Loss computation
Backward pass
Optimizer step
Zeroing gradients
8.2 Training Mode vs Evaluation Mode
model.train()
model.eval()
Why mode matters for dropout and batch norm
8.3 Tracking Training Progress
Loss curves
Accuracy curves
Validation performance
Overfitting indicators
8.4 Saving and Loading Models
state_dict
Saving model weights
Loading model weights
Saving checkpoints
9. Datasets and DataLoaders
9.1 PyTorch Dataset API
Dataset
Custom datasets
__len__
__getitem__
9.2 DataLoader
Batching
Shuffling
Parallel loading
num_workers
9.3 Data Splitting
Training set
Validation set
Test set
Data leakage
9.4 Data Preprocessing
Normalization
Standardization
Encoding labels
Transform pipelines
10. Model Evaluation
10.1 Evaluation Metrics
Accuracy
Precision
Recall
F1 score
Confusion matrix
ROC-AUC
10.2 Evaluation Workflow
Disabling gradients
Batch evaluation
Averaging losses
Interpreting predictions
10.3 Error Analysis
Misclassified examples
Class imbalance
Bias in predictions
Failure patterns
11. Overfitting and Regularization
11.1 Underfitting vs Overfitting
Bias-variance tradeoff
Training loss vs validation loss
11.2 Regularization Techniques
Weight decay
Dropout
Early stopping
Data augmentation
Label smoothing
11.3 Improving Generalization
More data
Simpler models
Better features
Better validation strategy
12. Neural Network Architectures
12.1 Feedforward Neural Networks
Multilayer perceptrons
Fully connected networks
Use cases
12.2 Convolutional Neural Networks
Convolutions
Filters and kernels
Padding and stride
Pooling
CNNs for images
12.3 Recurrent Neural Networks
Sequential data
RNNs
LSTMs
GRUs
12.4 Attention and Transformers
Attention mechanism
Self-attention
Positional encoding
Transformer encoder/decoder idea
13. Practical PyTorch Workflow
13.1 Standard Project Structure
Data
Model
Training script
Evaluation script
Configs
13.2 Debugging PyTorch Models
Shape errors
Device errors
Loss not decreasing
NaN losses
Gradient issues
13.3 Best Practices
Start with a small model
Overfit a tiny batch first
Track tensor shapes
Use validation properly
Save checkpoints
14. Advanced Training Concepts
14.1 Initialization
Why initialization matters
Xavier initialization
He initialization
PyTorch default initialization
14.2 Normalization
Batch normalization
Layer normalization
When normalization helps
14.3 Gradient Techniques
Gradient clipping
Gradient accumulation
Mixed precision training
14.4 Transfer Learning
Pretrained models
Feature extraction
Fine-tuning
Freezing and unfreezing layers
15. Deployment and Inference Basics
15.1 Inference Workflow
Loading a trained model
Preprocessing input
Running prediction
Postprocessing output
15.2 Model Export
TorchScript
ONNX basics
When export is useful
15.3 Performance Considerations
Batch inference
GPU inference
Model size
Latency vs accuracy
16. Suggested Mini Projects
16.1 Regression Model from Scratch
Predict continuous values using a simple neural network
16.2 Binary Classification
Build a classifier using BCE loss
16.3 Multiclass Classification
Train on MNIST or Fashion-MNIST
16.4 Image Classification with CNNs
Build a convolutional network
16.5 Transfer Learning Project
Fine-tune a pretrained vision model